scripts/wasm-strip.js

```diff
@@ -0,0 +1,363 @@
+            return;
+          }
+          const dataView = new DataView(buff);
+          const sectionType = dataView.getUint8(relativeOffset);
+          const [size, sizeOfSize] = readULeb128(dataView,
+                                                 relativeOffset + 1);
+
+          const sectionInfo = { sectionType,
+                                offset: currOffset,
+                                contentSize: size,
+                                length: size + 1 + sizeOfSize };
+          currOffset += 1 + sizeOfSize; // Section id + size
+
+          if (sectionType === CUSTOM_SECTION_ID) {
+            skipCurrentSection = true;
+            result.removedSections.push(sectionInfo);
+          } else {
+            skipCurrentSection = false;
+            writeStream.write(Buffer.from(buff.slice(relativeOffset,
+                                                     currOffset - chunkBaseOffset)));
+            result.keptSections.push(sectionInfo);
+          }
+          nextSectionOffset = size + currOffset;
+        }
+        const newOffset = Math.min(maxOffsetInChunk, nextSectionOffset);
+        if (!skipCurrentSection) {
+          writeStream.write(Buffer.from(buff.slice(currOffset - chunkBaseOffset,
+                                                   newOffset - chunkBaseOffset)));
+        }
+        currOffset = newOffset;
+      }
+    }
+  });
+}
+
+function checkMagicNumberAndVersion(buff) {
+  if (buff.byteLength < 8) {
+    throw new Error("Error: Not a valid WebAssembly file: file too short");
+  }
+  const dataView = new DataView(buff);
+  if (dataView.getUint32(0) !== WASM_MAGIC_NUMBER) {
+    throw new Error("Error: Not a valid WebAssembly file: no magic number");
+  }
+  if (dataView.getUint32(4, true) !== 1) {
+    throw new Error("Error: Unsupported WebAssembly version: ",
+                    dataView.getUint32(4, true));
+  }
+}
+
+function concatArrayBuffers(...buffers) {
+  const ret = new Uint8Array(buffers.reduce((acc, buff) => acc + buff.byteLength, 0));
+  let currOffset = 0;
+  for (let i = 0; i < buffers.length; i++) {
+    const buffer = buffers[i];
+    ret.set(buffer, currOffset);
+    currOffset += buffer.byteLength;
+  }
+  return ret.buffer;
+}
+
+/**
+ * WebAssembly got that weird "LEB128" integer format.
+ *
+ * This function parses the unsigned version into a JS number.
+ * @param {DataView} dv
+ * @param {number} offset - Offset, in bytes, in the buffer where that number
+ * starts.
+ * @returns {Array.<number>}
+ */
+function readULeb128(dv, offset) {
+  let res = 0;
+  let currShift = 0;
+  let currOffset = offset;
+  let nbSize = 0;
+  while (currOffset < dv.byteLength) {
+    // take next bytesize=0x00000070
+    const byte = dv.getInt8(currOffset);
+
+    // do some shenanigans to extract the interesting value.
+    // (all but the highest bit, which just tells us when this is the end).
+    const val = byte & 0x7F;
+
+    res |= (val << currShift);
```

Comment:
That part looks very interesting and a little bit harder to follow, I think I understand the AND operator on Bit and the Bit shift to left.

But I don't understand why is it done? for what purpose?

Reply:
> I think I understand the AND operator

It's an OR :p

It's a shift + OR to re-constuct the original integer. In LEB128 integers are splitted into 7-bit groups (as the highest bit of each byte indicate length) so we have to reconstruct it in 8-bit groups somehow.

The idea is to remove the highest bit and to reconstruct the final integer. This is done by:

  1. shifting, if necessary by the right multiple of 7 (0, then 7, then 14 and so on...)
 
  2. "combining" it into the final number, by using a bitwise OR

---

For example, if you have: `11000001.00111110` 
  1. We will first consider the first byte, `11000001`
  2. `const val = byte & 0x7F` (0b01111111) allows to not consider the highest bit for now as it is only useful for obtaining the length: `0b1000001`
  3. then we will shift it by 0 (0 * 7) and "OR it" to `res`, which is initially set to `0`: `0b00000000  | 0b1000001 === 0b01000001`
  4. we check the higher bit by doing `& 0x80` (`0b10000000`), here it's `0b10000000`,  which is not 0, so it's not the end of the number.
  
  5. We consider the second byte, `00111110`
  6. like 2., we remove the higher bit: `0111110`
  7. We shift it by 7 (1 * 7) (`0b01111100000000`) and "OR it" to `res`: `0b01000001 | 0b01111100000000 === 0b01111101000001` (note: this has for effect to  overwrite the 8th bit of the previous `res` value, but we don't care, as it is always set to 0).
  8. The higher bit is `0`, we've reached the end of the number

That number is now `0b01111101000001`, which if separated in 7-bit groups make:
`0111110.1000001`

This is equivalent to the input (`11000001.00111110`) but reversed because it's little endian and with the most significant bit of each byte removed (because it is only there to communicate length).